File: dart/glue/lib/src/lib/bool/and.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

final Ir and_ = IrSpecial(andImpl);

Eval<Ir> andImpl(List<Ir> args) {
  return switch (args) {
    [final a, final b] => eval(a).bind((a_) {
      return isFalsy(a_)
          ? Eval.pure(IrBool(false))
          : eval(b).bind((v) => Eval.pure(IrBool(isTruthy(v))));
    }),
    _ => throwError(wrongNumberOfArguments()),
  };
}

File: haskell/glue/src/Glue/Lib/Bool/And.hs
module Glue.Lib.Bool.And where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception (wrongNumberOfArguments)
import Glue.IR (IR (..), isFalsy, isTruthy)

and_ :: IR Eval
and_ = Special andImpl

andImpl :: [IR Eval] -> Eval (IR Eval)
andImpl [a, b] = do
    a' <- eval a
    if isFalsy a'
        then pure (Bool False)
        else Bool . isTruthy <$> eval b
andImpl _ = throwError wrongNumberOfArguments
